pandas/core/accessor.py

```diff
@@ -129,3 +102,136 @@ def f(self, *args, **kwargs):
             # don't overwrite existing methods/properties
             if overwrite or not hasattr(cls, name):
                 setattr(cls, name, f)
+
+
+# Ported with modifications from xarray
+# https://github.com/pydata/xarray/blob/master/xarray/core/extensions.py
+# 1. We don't need to catch and re-raise AttributeErrors as RuntimeErrors
+
+
+class _CachedAccssor(object):
+    """Custom property-like object (descriptor) for caching accessors.
+
+    Parameters
+    ----------
+    name : str
+        The namespace this will be accessed under, e.g. ``df.foo``
+    accessor : cls
+        The class with the extension methods. The class' __init__ method
+        should expect one of a ``Series``, ``DataFrame`` or ``Index`` as
+        the single argument ``data``
+    """
+    def __init__(self, name, accessor):
+        self._name = name
+        self._accessor = accessor
+
+    def __get__(self, obj, cls):
+        if obj is None:
+            # we're accessing the attribute of the class, i.e., Dataset.geo
+            return self._accessor
+        accessor_obj = self._accessor(obj)
+        # Replace the property with the accessor object. Inspired by:
+        # http://www.pydanny.com/cached-property.html
+        # We need to use object.__setattr__ because we overwrite __setattr__ on
+        # NDFrame
+        object.__setattr__(obj, self._name, accessor_obj)
+        return accessor_obj
+
```

Comment:
it might make sense to have a few methods here which are AbstractMethodError to guide the api, eg. ``validate``

Reply:
I'd prefer to leave that open to the person implementing the accessor. Perhaps for some accessors, there's no need for validation.

Reply:
this is too general we need to be opionated, if there is no validation then that should be explicit.

Reply:
Not sure what you mean be "too general". The purpose of this is to give an officially blessed way to do "DataFrame.foo = MyAccessor"; what you do past that point is up to the library. It's not like `CachedAccessor` ever has a chance to see the data so that it can validate things. 

The only restriction on the accessor is that its init method be `__init__(self, pandas_obj)`. I'll document that, but I'm not sure what else there is to be done.